### PR TITLE
feat: use new types from `@octokit/webhooks-definitions`

### LINKED ISSUE #364: `pull_request.merged` event listener never invoked
**What happened?**

`webhooks.on('pull_request.merged', ...)` is never invoked. Instead, the listener for `pull_request.closed` is called.

**What did you expect to happen?**

I expected the listener attached to `pull_request.merged` be called when a pull request is merged.

**What the problem might be**

`merged` isn't even a valid action on the `pull_request` payload. Instead, a pull request is merged if its `action === 'closed'` and `merged === true`. 

Either a special case should be added to get `pull_request.merged` to work, or it should be removed to avoid this bug.

### PR DESCRIPTION
**Pull request by G-Rath**
Here it is 🎉 

There's a few things to polish off and discuss before landing this, but the bulk of it is done so making the PR now.

* we're missing some events in our schema definitions.
* I've done a big rename of the types so that they're more accurate (have exported the renamed types under their old names too to be less breaking).
* I've improved the types in places so that they're accurate
* have refactored the type generation script into TS - happy to move it back to JS, or to convert the other scripts into TS
  * I tried originally trying to build the type via types, but it's just too hard (if at all possible) - you can get close with 4.1 features, but even then it's a lot of work for TS to do meaning IDE autocompletion suffers vs us just generating a static type
* probably something else I've forgotten to mention 😅 

I'm breaking some things out into their own little PRs where possible.

### CODE REVIEW

**Review by wolfy1339 — DISMISSED** [review]

**Comment by G-Rath** [comment]
I think we're good to go - note this'll be rebased at least twice when I merge in #403 & #410, so approves won't stick until then

While I've marked everything in the old generated payloads as `deprecated`, I'd like to delete them asap.
Also, there's probably a couple of redundant type aliases in there that could be cleaned up, but shouldn't be a big deal if that happens later.

**Review by wolfy1339 — DISMISSED** [review]

**Review by oscard0m — COMMENTED** [review]
I don't feel confident enough to give an approval but, for what I read, everything looks good to me :)

File: src/generated/webhook-names.ts [hunk 1]
```diff
@@ -41,8 +41,8 @@ export const webhookNames = [
   "installation.created",
   "installation.deleted",
   "installation.new_permissions_accepted",
-  "installation.suspend",
-  "installation.unsuspend",
+  "installation.suspended",
```
Comment (wolfy1339) [inline src/generated/webhook-names.ts:44]: These values are incorrect, I fixed these in octokit/webhooks#300

File: src/generated/webhook-names.ts [hunk 2]
```diff
@@ -41,8 +41,8 @@ export const webhookNames = [
...
+  "installation.unsuspended",
```
Comment (wolfy1339) [inline src/generated/webhook-names.ts]: These values are incorrect, I fixed these in [octokit/webhooks#300](https://github.com/octokit/webhooks/pull/300)
https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads#installation

File: src/generated/get-webhook-payload-type-from-event.ts [hunk 3]
```diff
@@ -1,215 +1,412 @@
...
+  fork: ForkEvent;
+  github_app_authorization: GithubAppAuthorizationEvent;
+  "github_app_authorization.revoked": GithubAppAuthorizationRevokedEvent;
+  gollum: GollumEvent;
+  installation: InstallationEvent;
+  "installation.created": InstallationCreatedEvent;
+  "installation.deleted": InstallationDeletedEvent;
+  "installation.new_permissions_accepted": InstallationNewPermissionsAcceptedEvent;
+  "installation.suspended": InstallationSuspendedEvent;
+  "installation.unsuspended": InstallationUnsuspendedEvent;
```
Comment (wolfy1339) [inline src/generated/get-webhook-payload-type-from-event.ts]: These values are incorrect, I fixed these in octokit/webhooks#300
https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads#installation

**Review by wolfy1339 — CHANGES_REQUESTED** [review]

**Comment by gr2m** [comment]
> While I've marked everything in the old generated payloads as `deprecated`, I'd like to delete them asap.

Should these deprecations be part of this PR or did they already happen?

`@octokit/webhooks` has several deprecated APIs, we can cut a breaking release anytime and get rid of them all at once.

**Comment by G-Rath** [comment]
> Should these deprecations be part of this PR or did they already happen?

I've done them in this PR, but can easily break them out into their own PR if you like. I figured it made sense to deprecate in the same PR that lands the replacement.

File: src/generated/webhook-names.ts (hunk 2 above)
Comment (wolfy1339) [reply src/generated/webhook-names.ts]: Fixed in https://github.com/octokit/webhooks/pull/324

File: src/generated/get-webhook-payload-type-from-event.ts (hunk 3 above)
Comment (wolfy1339) [reply src/generated/get-webhook-payload-type-from-event.ts]: Fixed in https://github.com/octokit/webhooks/pull/324

**Review by wolfy1339 — DISMISSED** [review]
Once you update `@octokit/webhooks-definitions` with the required patch it will be good to merge.
Otherwise LGTM

**Comment by G-Rath** [comment]
I've pinned the version constraint for `@octokit/webhooks-definitions` to be an exact version, and updated the `update` workflow to update the constraint with the latest version.

The reason for this is that otherwise it can be updated by anything that regenerates the lockfile, which we've got renovate doing every week, but we need to make sure `generate-types` is run every time the version of this package changes which won't happen so it's PRs will always fail CI.

**Review by wolfy1339 — DISMISSED** [review]

File: .github/workflows/update.yml [hunk 4]
```diff
@@ -11,8 +11,7 @@ jobs:
       - uses: actions/checkout@v2
       - uses: actions/setup-node@v2
       - run: npm ci
-      - run: npm install @octokit/webhooks-definitions@latest
-      - run: npm run update-known-events
+      - run: npm install -ED @octokit/webhooks-definitions@latest
```
Comment (gr2m) [inline .github/workflows/update.yml]: I'd prefer not to use abbreviations. They are great for typing commands in your terminal, but they are harder to read/understand to others and your future self. Can you use the full flag names instead?

File: package.json [hunk 5]
```diff
@@ -12,12 +12,11 @@
   "scripts": {
     "build": "pika build",
     "coverage": "jest --coverage && open coverage/lcov-report/index.html",
-    "generate-types": "node scripts/generate-types.js",
+    "generate-types": "ts-node -T scripts/generate-types.ts",
```
Comment (gr2m) [inline package.json]: I guess `-T` is `--transpile-only`? Can you write that out please?

File: src/event-handler/remove-listener.ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
-import { State } from "../types";
+import { EmitterEventName, State } from "../types";
 
 export function removeListener(
   state: State,
-  webhookNameOrNames: string | string[],
+  webhookNameOrNames: EmitterEventName | EmitterEventName[],
```
Comment (gr2m) [inline src/event-handler/remove-listener.ts:5]: Will this work with `*` and `error`? Related: https://github.com/octokit/webhooks.js/pull/416#pullrequestreview-579913781

**Review by gr2m — COMMENTED** [review]
My understanding is that we can merge this in as a feature release, as it does not introduce any breaking changes? If so, +1 for merging from me. Otherwise please let me know what the breaking changes are. If breaking changes are TypeScript types only we can release a feature release, I don't consider TypeScript type changes breaking changes as they don't affect production code, they only occur at build time

File: src/event-handler/remove-listener.ts (hunk 6 above)
Comment (G-Rath) [reply src/event-handler/remove-listener.ts:5]: Yup! `EmitterEventName` is the `keyof EmitterEventMap`, which is of type `EmitterWebhookEventMap & { error: WebhookEventHandlerError; };`

image

**Review by gr2m — DISMISSED** [review]

**Review by wolfy1339 — DISMISSED** [review]

**Review by wolfy1339 — APPROVED** [review]

**Comment by G-Rath** [comment]
Alright, let's do this! we can fix up any issues in post 😎

**Comment by G-Rath** [comment]
> Error: src/event-handler/index.ts(10,17): error TS4058: Return type of exported function has or is using name 'BaseWebhookEvent' from external module "/home/runner/work/webhooks.js/webhooks.js/src/types" but cannot be named.

well, that's odd. I'll see if I can figure out what's going wrong.

**Comment by G-Rath** [comment]
I know why this is breaking - will have a PR up shortly

**Pull request merged**